launcher.py

```diff
@@ -0,0 +1,206 @@
+    logging.info(f"Using tags {tags} to find matching image.")
+
+    manifest_digest = get_manifest_digest(registry, image_name, tags, image_digest)
+
+    name_and_digest = image_name + "@" + manifest_digest
+
+    proc = run(["docker", "pull", name_and_digest])
+
+    if proc.returncode:
+        raise RuntimeError("docker pull returned non-zero exit code %d" % proc.returncode)
+
+    proc = run(["docker", "image", "inspect", "--format", "{{index .ID}}", name_and_digest], capture_output=True)
+
+    if proc.returncode:
+        raise RuntimeError("docker image inspect returned non-zero exit code: %d" % proc.returncode)
+
+    pulled_image_digest = proc.stdout.decode('utf-8').strip()
+
+    if pulled_image_digest != image_digest:
+        raise RuntimeError("Wrong image digest %s. Expected digest is %s" % (pulled_image_digest, image_digest))
+
+    # Generate a quote before extending RTMR3 with the image digest
+    proc = run(['curl', '--unix-socket', '/var/run/dstack.sock', '-X', 'POST',
+                'http://dstack/GetQuote', '-H', 'Content-Type: application/json', '-d',
+                '{"report_data": ""}'], capture_output=True)
+
+    if proc.returncode:
+        raise RuntimeError("getting quote failed with error code %d" % proc.returncode)
+    logging.info("Quote: %s" % proc.stdout.decode('utf-8').strip())
+
+    # Python's requests package cannot natively talk HTTP over a unix socket (which is the API
+    # exposed by dstack's guest agent). To avoid installing another Python depdendency, namely
+    # requests-unixsocket, we just use curl.
+    extend_rtmr3_json = '{"event": "launcher-image-digest","payload": "%s"}' % image_digest.split(':')[1]
+    proc = run(['curl', '--unix-socket', '/var/run/dstack.sock', '-X', 'POST',
+                'http://dstack/EmitEvent', '-H', 'Content-Type: application/json', '-d',
+                extend_rtmr3_json])
+
+    if proc.returncode:
+        raise RuntimeError("extending rtmr3 failed with error code %d" % proc.returncode)
+
+    # Get quote after extending RTMR3 with the image digest
+    proc = run(['curl', '--unix-socket', '/var/run/dstack.sock', '-X', 'POST',
+                'http://dstack/GetQuote', '-H', 'Content-Type: application/json', '-d',
+                '{"report_data": ""}'], capture_output=True)
+
+    if proc.returncode:
+        raise RuntimeError("getting quote failed with error code %d" % proc.returncode)
+
+    logging.info("Quote: %s" % proc.stdout.decode('utf-8').strip())
+
+    # Build the docker command we use to start the app, i.e., mpc node
+    docker_cmd = ['docker', 'run']
+
+    for env_file in ['/tapp/.host-shared/.user-config']:
+        if os.path.isfile(env_file):
+            docker_cmd += ['--env-file', env_file]
+
+    # Forward ports from container to CVM
+    # This port is the MPC node port. It must be forwarded and cannot be remapped.
+    # TODO Adjust and/or make configurable according to `port_override` from the mpc node.
```

Comment:
I assume this isn't critical for now. But would be good to have a follow-up to track it.